submissions/vovan-zt/Frogger-Game/app.js

```diff
@@ -0,0 +1,108 @@
+const gridSize = {
+    width: 101,
+    height: 81,
+};
+
+const fieldSize = {
+width: gridSize.width*5,
+height: gridSize.height*6,
+top:0,
+left:0,
+};
+
+const startPlayerPosition = {
+x:gridSize.width*2 ,
+y:gridSize.height*5,
+};
+
+const enemySpeed = 200;
+const  Enemy = function(x, y, speed) {  
+    this.x = x;
+    this.y = y;
+    this.speed = speed;
+    this.sprite = 'images/enemy-bug.png';
+};
+
+Enemy.prototype.update = function(dt) {
+    this.x += this.speed*dt;
+
+    if (this.x > fieldSize.width) {
+        this.x = -gridSize.width;
+        this.speed = 100 + Math.floor(Math.random() * enemySpeed);
+    };
+
+    if (player.x + gridSize.width > this.x && player.x < this.x + gridSize.width &&
+        player.y + gridSize.height > this.y && player.y < this.y + gridSize.height) {
+        player.x = startPlayerPosition.x;
+        player.y = startPlayerPosition.y;
+    }  
+};
+
+Enemy.prototype.render = function() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+const  Player = function(x, y) {
+    this.x = x;
+    this.y = y;
+    this.sprite = 'images/char-boy.png';
+};
+
+Player.prototype.update = function (dt) {
+ 
+};
+
+Player.prototype.render = function () {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Player.prototype.handleInput = function (keyPress) {
+    if (keyPress == 'up' && this.y > fieldSize.top) {
+        this.y -= gridSize.height;
+    }
+
+    if (keyPress == 'down' && this.y < fieldSize.width-gridSize.width) {
+        this.y += gridSize.height;
+    }
+
+    if (keyPress == 'left' && this.x > fieldSize.left) {
+        this.x -= gridSize.width;
+    }
+
+    if (keyPress == 'right' && this.x < fieldSize.width-gridSize.width) {
+        this.x += gridSize.width;
+    }
+
+    if (this.y <= fieldSize.top) {
+        setTimeout(() => {
+            this.x = startPlayerPosition.x;
+            this.y = startPlayerPosition.y;
+        }, 600);
+    };
+}; 
+
+let positionEnemies= [];
```

Comment:
Do we need this being global?

Reply:
In this case yes

Reply:
Yes, at some point I got why you need it global.
The proper pattern for recursive calls would be that every function call embeds an array received from a recurrent call into its own returning value.
Example: Build a list of items where as long as there are more than five items we add another list shorter by 5 than original, and so on until less than 5 items added.
_Doesn't really make sense and not helpful in your case. Just to illustrate how a function can populate an array without global variables._
```
function populateList(numberOfItems, nestingLevel = 1) {
  let result = [];
  for (let i = 0; i < numberOfItems; i++) result.push(`${nestingLevel}-${i+1}`);
  if (numberOfItems > 5) {
    result = [...result, ...populateList(numberOfItems - 5, nestingLevel + 1)];
  }
  return result;
}

console.log(populateList(18));
```

An alternative version, where you may not even need a variable at all:
```
function populateList2(numberOfItems, nestingLevel = 1) {
  return [
    ...Array.from({length: numberOfItems}, (_, i) => `${nestingLevel}-${i+1}`),
    ...(numberOfItems > 5 ? populateList(numberOfItems - 5, ++nestingLevel) : []),
  ];
}

console.log(populateList(18));
```

When spreads are not a good approach for readability reasons, `Array#flat` is to the rescue.
```
[[1,2,3], [4,5,6], [7,8,9]].flat() // [1, 2, 3, 4, 5, 6, 7, 8, 9]
```

Hope the above helps.

Reply:
Also try to describe your task in plain language. Like "I want to have as a result an array where...".
Quite often there are programming language constructions that reflect task requirements nearly 1:1.

Reply:
![TEST](https://user-images.githubusercontent.com/75982068/187035948-49e79a8e-5af4-4f22-8e7c-9a93eb7deb49.png)


I just wanted to get such an array, with the height coordinates where the enemies should appear.

Reply:
Frankly, I do not fully understand how this will help me here.